Packet.getlayer() method can work with string layer name

vorotnikovdk opened the pull request:
I found that Packet.getlayer() can't work when _cls_ param is string. For example, Packet.haslayer() can do it.
I created the fix, review it, please

p-l- commented:
Hi,

Can you explain what was broken before this PR and is fixed now? Thanks

vorotnikovdk commented:
Sure.
Before PR:
```
In [1]: from scapy.all import *
WARNING: No route found for IPv6 destination :: (no default route?)

In [2]: pkt = Ether()/Dot1Q()/IP()/UDP()

In [3]: pkt.getlayer(Dot1Q) # this was ok
Out[3]: <Dot1Q  type=0x800 |<IP  frag=0 proto=udp |<UDP  |>>>

In [4]: pkt.getlayer("Dot1Q") # this didnt work, None was returned

In [5]: 
```

After PR:
```
In [1]: from scapy.all import *
WARNING: No route found for IPv6 destination :: (no default route?)

In [2]: pkt = Ether()/Dot1Q()/IP()/UDP()

In [3]: pkt.getlayer(Dot1Q) # this is still ok
Out[3]: <Dot1Q  type=0x800 |<IP  frag=0 proto=udp |<UDP  |>>>

In [4]: pkt.getlayer("Dot1Q") # this works too
Out[4]: <Dot1Q  type=0x800 |<IP  frag=0 proto=udp |<UDP  |>>>

In [5]: 
```

I think that getting layer by its string name can be usefull

p-l- commented:
I understand this might be confusing but it is the expected behavior. There is no reason to accept the layer name rather than the layer in `.haslayer()`.

vorotnikovdk commented:
Sorry for writing in closed PR, can you please explain some things?

It is really confusing because of this code:
```
if type(cls) is str and "." in cls:
    ccls,fld = cls.split(".",1)
```
Looks line scapy is expexting string in this method. If _cls_ can be only layer, but not layer name, this code is strange.

Moreover.
String `self.__class__.name == ccls` is comparing _ccls_ (which is a class or a string after code I mentioned above) with descriptor created by python because of using __slots__.
```
print type(self.__class__.name) # ==> <type 'member_descriptor'>
```
It is confusing too. Is it expected? Or it is side effect of using __slots__? This is second strange thing.

Third:
I checked version 2.3.2 without __slots__.
In this version `self.__class__.name == ccls` is comparing _ccls_ with class attribute _name_.
So this `pkt.getlayer("UDP")` works.
And `pkt.getlayer("TCP")` works.
And this `pkt.getlayer("Ether")` doesn't work. 
But `pkt.getlayer("Ethernet")` works because of `Ether.name == "Ethernet"`
And this `pkt.getlayer("Dot1Q")` also doesn't work.
And if I use Dot1Q.name `pkt.getlayer("802.1Q")` it doesnt work because of `cls.split(".",1)`

It looks like before __slots__ I could get layers by its class attribute _name_ (but there were a problem for classes which name contains a dot). I have no other explanations to checking `if type(cls) is str`.
And after adding __slots__ this possibility was broken completly.

Am I wrong?

If there were support for getting layers by string name, I think it will be good to turn it back and make it the same with haslayer() - using class's name such "Dot1Q" but not using class attribute _name_ such "Ethernet" or "802.1Q"

p-l- commented:
First, it's totally OK to write on a closed PR to continue the discussion.

Second, you're totally right. So thanks for this!

vorotnikovdk commented:
One more idea.
Some old code can use class attribute "name" for getting layer. It is broken in current code and it will be still broken after my PR.
I can add one more check to keep compaitability with such code.
So all this variants:
`pkt.getlayer(Ether) # layer`
`pkt.getlayer("Ether") # layer name`
`pkt.getlayer("Ethernet") # attribute "name"`
will work.
Should I do it? Or using attribute "name" is wrong and all scapy users must avoid it?

vorotnikovdk commented:
Maybe idea is bad
haslayer() can't use attribute _name_.
I think it will be good if getlayer and haslayer will work in unified style: layer or layername.
Attribute _name_ has problems with names like 802.3 and 802.1Q

p-l- commented:
I think it's best if `.haslayer()` and `.getlayer()` share the same logic. What do you think?

vorotnikovdk commented:
I am agree.
